### PR TITLE
Add support for updating secrets

### PR DESCRIPTION
**Pull request by wallyworld**
Add support for updating secrets:
- secrets-update hook command
- api and apiserver facade methods
- "juju" services service methods
- mongo / state methods

The hook command can be used to update either the secret data, the rotate interval, or both.

Also tweak error landing in the list-secrets CLI so errors are displayed in YAML/JSON output.

## QA steps

bootstrap and deploy ubuntu charm

Create the secret
```plain
$ juju exec --unit ubuntu/0 "secret-create password foo=bar hello=world"
secret://v1/5a4f74fc-3b37-46d9-8aa0-611f7f838197/df931ce7-ce22-4ce5-849a-4d315c805832/ubuntu.password
$ juju exec --unit ubuntu/0 "secret-get secret://v1/ubuntu.password"
foo: bar
hello: world

$ juju secrets
ID  Revision  Rotate  Backend  Path             Age
1          1  never   juju     ubuntu.password  33 seconds ago  
```

Update the rotate interval
```
$ juju exec --unit ubuntu/0 "secret-update password --rotate=1h"
secret://v1/5a4f74fc-3b37-46d9-8aa0-611f7f838197/df931ce7-ce22-4ce5-849a-4d315c805832/ubuntu.password?revision=1
$ juju secrets
ID  Revision  Rotate  Backend  Path             Age
1          1  1h      juju     ubuntu.password  8 seconds ago  

$ juju secrets --format yaml --show-secrets
- ID: 1
  URL: secret://v1/ubuntu.password
  revision: 1
  path: ubuntu.password
  rotate-interval: 1h0m0s
  version: 1
  backend: juju
  create-time: 2021-08-24T03:22:07Z
  update-time: 2021-08-24T03:23:43Z
  value:
    foo: bar
    hello: world
```

Update the value
```
$ juju exec --unit ubuntu/0 "secret-update password foo2=bar"
secret://v1/5a4f74fc-3b37-46d9-8aa0-611f7f838197/df931ce7-ce22-4ce5-849a-4d315c805832/ubuntu.password?revision=2
$ juju secrets --format yaml --show-secrets
- ID: 1
  URL: secret://v1/ubuntu.password
  revision: 2
  path: ubuntu.password
  rotate-interval: 1h0m0s
  version: 1
  backend: juju
  create-time: 2021-08-24T03:22:07Z
  update-time: 2021-08-24T03:24:46Z
  value:
    foo2: bar
```

Set rotate interval back to "never"
```
$ juju exec --unit ubuntu/0 "secret-update password --rotate=0"
secret://v1/5a4f74fc-3b37-46d9-8aa0-611f7f838197/df931ce7-ce22-4ce5-849a-4d315c805832/ubuntu.password?revision=2
$ juju secrets --format yaml --show-secrets
- ID: 1
  URL: secret://v1/ubuntu.password
  revision: 1
  path: ubuntu.password
  version: 1
  backend: juju
  create-time: 2021-08-24T03:22:07Z
  update-time: 2021-08-24T03:25:25Z
  value:
    foo2: bar

$ juju secrets
ID  Revision  Rotate  Backend  Path             Age
1          3  never   juju     ubuntu.password  11 seconds ago  
```

### CODE REVIEW

**Comment by wallyworld** [comment]
[update secrets support](https://trello.com/c/NMZ4AAoZ/1726-update-secrets-support)

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 1]
```diff
@@ -81,6 +81,47 @@ func (s *SecretsManagerAPI) createSecret(ctx context.Context, arg params.CreateS
...
+	if URL.Attribute != "" {
+		return "", errors.NotSupportedf("updating a single secret attribute %q", URL.Attribute)
+	}
+	if URL.Revision > 0 {
+		return "", errors.NotSupportedf("updating secret revision %d", URL.Revision)
+	}
+	if arg.RotateInterval < 0 && len(arg.Data) == 0 {
+		return "", errors.New("either rotate interval or data must be specified")
+	}
+	URL.ControllerUUID = s.controllerUUID
```
Comment (ycliuhw) [inline apiserver/facades/agent/secretsmanager/secrets.go:112]: do we wanna reject if the the URL does provide a wrong `ControllerUUID` and/or `ModelUUID`?

File: core/secrets/secret.go [hunk 2]
```diff
@@ -195,7 +194,7 @@ func (u *URL) String() string {
 	}
 	fullPath = append(fullPath, u.Path)
 	str := strings.Join(fullPath, "/")
-	if u.Revision > 1 {
+	if u.Revision > 0 {
```
Comment (ycliuhw) [inline core/secrets/secret.go:197]: Probably we could define this URL like `url.URL{Scheme: "secret"}`

File: state/secrets.go [hunk 3]
```diff
@@ -96,16 +97,23 @@ type secretsStore struct {
...
-		return nil, errors.Trace(err)
+		return nil, nil, errors.Trace(err)
+	}
+	interval := p.RotateInterval
+	if interval < 0 {
+		interval = 0
 	}
-	return &secretMetadataDoc{
+	URL := secrets.NewSimpleURL(p.Version, p.Path)
+	URL.ControllerUUID = p.ControllerUUID
```
Comment (ycliuhw) [inline state/secrets.go:110]: It seems we do this in both facade and state.
The URL here is like the ID, it's identical for each revision of the secret. 
I would prefer to generate the URL from the facade, then always use that one for the whole process.

**Review by ycliuhw — APPROVED** [review]
LGTM, just a few suggestions.

File: state/secrets.go (hunk 3 above)
Comment (wallyworld) [reply state/secrets.go]: ok

File: core/secrets/secret.go (hunk 2 above)
Comment (wallyworld) [reply core/secrets/secret.go]: ok

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:119]: yeah, done

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**